desi_tile_redshifts

sbailey opened the pull request:
@akremin this PR provides `desi_tile_redshifts` as an alternative to `desi_nightly_redshifts` (which I haven't deleted yet).  Instead of running all redshifts in a single job, this generates one batch script per tile-group, with filtering options for whether that group is all exposures on a single night, across multiple nights, from a separately curated external list, etc.  Examples:
```
All exposures of tile 80605 on night 20201215:

    desi_tile_redshifts --tile 80605 --night 20201215

Exposures E1 E2 E3 on night 20201215 (auto splitting by TILEID if needed):

    desi_tile_redshifts --night 20201215 --expid 67972 67973 67968 67969

Tile 80605 combined across multiple nights:

    desi_tile_redshifts --tile 80605 --night 20201214 20201215 --group blat

Tile 80605 combined across all nights:

    desi_tile_redshifts --tile 80605 --group all

Generate scripts for every tile on 20201215 but don't submit batch jobs:

    desi_tile_redshifts --night 20201215 --nosubmit

Use exposures from a separately curated input list:

    desi_tile_redshifts --explist explist-deep.txt --group deep

Not supported yet: multiple tiles on a single night in a single call (call each individually instead):

    desi_tile_redshifts --night 20201215 --tileid 80605 80606 80607
```

Running `desi_tile_redshifts --night YEARMMDD` will submit jobs for all tiles observed on night `YEARMMDD`, equivalent to what we currently do with `desi_nightly_redshifts --batch YEARMMDD` but splitting them into separate jobs to provide greater chance of individual success without timeout, and making it easier to re-submit individual tiles if needed after cleanup.  One tile-night across 10 spectrographs takes <5 min per job.

For consideration: These generate scripts and logs in `$DESI_SPECTRO_REDUX/$SPECPROD/run/scripts/tiles/TILEID/GROUP`, for symmetry with the output files in `$DESI_SPECTRO_REDUX/$SPECPROD/tiles/TILEID/GROUP`, but note that the scripts/logs are different from cascades which had nightly redshift scripts in `run/script/night/NIGHT/` and other tile groupings in `run/script/tiles/GROUP/TILEID`.

Features this does *not* yet provide:
  * grouping multiple tiles into a larger >10 node job; a future update could take the scripts generated here and source them from within a larger job.
  * filtering by spectrograph number or CAMWORD/BADCAM columns, though there are some hooks in place for that and it at least will quickly skip over any missing inputs
  * hooks into the pipeline to run `desi_tile_redshifts --night N --tile T` automaticlly after the poststdstar jobs have run, with slurm dependencies to wait for those to finish finish first.
  * hooks into the pipeline to run `desi_tile_redshifts --tile T1 T2 T3 ... --group all` to update the multi-night redshifts for every tile that received new data tonight

akremin commented:
At first glance this looks excellent. I'll do a more thorough review tomorrow but I have one request if it's not too much effort:

I know you explicitly mention this isn't meant to support the pipeline, but it is practically there and it would be nice to get this in a state so that I can start putting the hooks in place on the pipeline side. (In fact, what you have would work for the pipeline, but in a rather inefficient way). Could the logic be set up such that if the night, tileid, and all expids are explicitly given then it doesn't need to generate your explist? If I'm reading the script correctly, the first thing it does is read in the exposure tables to generate an explist if one isn't provided, but it just uses the explist to derive those three bits of information.

sbailey commented:
Yes, that would be easy to add (tomorrow).  It would also be easy to add `--batch-afterok` and `--batch-afterany` options to support the pipeline if those would be immediately useful.

akremin commented:
You are welcome to add those additional options, but the pipeline defines the command line arguments to sbatch itself, so it won't use those or the already implemented reservation argument. I don't see any harm in having them for command line use, though.

sbailey commented:
@akremin I added the logic so that if `--tileid`, `--night`, and `--expid` are all specified, it will create its own exptable using those without reading from disk.

I also added a `--batch-dependency` option that gets passed to `sbatch --dependency` upon submission so that this can be used standalone, even if the regular pipeline would use `--nosubmit` to generate the script and then handle its own sbatch arguments upon submission (which makes sense).

If/when you/we use this within the automated pipeline, we could also refactor pieces into functions that can be imported and called instead of spawning a command line script, but for the initial pass I find it convenient to develop everything within a single self-contained script.

akremin commented:
The command line nature draws a clean parallel with the other portions of the code that make calls to desi_proc with a similar --nosubmit flag, so I'm happy with the way you have it setup.

I agree with you that down the line we could consider converting them all to the functional forms.

akremin commented on bin/desi_tile_redshifts:
```diff
@@ -0,0 +1,343 @@
+    reduxdir = desispec.io.specprod_root()
+    scriptdir = f'{reduxdir}/run/scripts/tiles/{tileid}/{group}'
+    os.makedirs(scriptdir, exist_ok=True)
+    batchscript = f'{scriptdir}/coadd-redshifts-{tileid}-{group}.slurm'
```
_If_ we want to support single exposures, they break this otherwise valid trend of `/{group}/coadd*-{group}.slurm`, `/{group}/spectra*-{group}.fits`, `/{group}/coadd*-{group}.fits`, `/{group}/zbest*-{group}.fits`, and  ` /{group}/redrock*-{group}.fits`.

I'll leave it up to you whether we want to have each saved under group={expid} and then consolidate all those directories post-factor into an /exposures/ directory. Or whether we want to complicate the logic here to allow for a change in the naming convention

akremin commented on bin/desi_tile_redshifts:
```diff
@@ -0,0 +1,343 @@
+    sys.exit(1)
+
+#- default group is the YEARMMDD night, but must be given if combining nights
+if args.group is None:
```
again if we want to support single exposures this could auto-detect a single exposure and name group={expid}. Or we could require the user to define it on the command line. Up to you, but wanted to mention it

akremin reviewed (commented):
This looks great. I have comments inline. Two are very nit-picky. The rest are things I wanted to mention and hear your opinion about, but I'll let you decide what action you want to take on them. Nothing here is a blocking factor. The script works as advertised.

I did one test (for my use case, because I'm biased):
`export SPECPROD=kremin_test`
`desi_tile_redshifts --night 20210214 --tile 80733 --expid 75873 75874 --nosubmit`

which worked properly. The job is waiting in the queue. I'd like to see it run but the script itself looks it's highly unlikely that it will fail.

I'll trust that you tested on the explist scenarios. So the only outstanding is whether we want to full support single exposures with this (see inline comment) or whether we use this as-is and move those files afterward.

sbailey commented on bin/desi_tile_redshifts:
```diff
@@ -0,0 +1,343 @@
+#- Generate the scripts and optionally submit them
+failed_jobs = list()
+for tileid in tileids:
+    ii = exptable['TILEID'] == tileid
```
I changed it in this case to have a clearer name, but as a heads up I use this kind of syntax a lot throughout the desispec code with the convention that single letters (e.g. i, j, k) are scalar indices and double letters (e.g. ii, jj, kk) are arrays of indices or arrays of booleans for slicing other arrays.

sbailey commented on bin/desi_tile_redshifts:
```diff
@@ -0,0 +1,343 @@
+
+    Args:
+        tileid (int): Tile ID
+        exptable (Table): has columns NIGHT EXPID to include
```
I added comments to clarify that.  This was partially motivated by the change of exposure_tables format during sv1 which tripped up stacking them across all nights.  Trimming down to just the columns needed and having the functions only require the minimal set of columns needed (as a generic Table) instead of the specific pipeline full exposures_table made this more robust to the exact input format (other columns can be there; they are just ignored)

akremin reviewed (approved):
I'm happy with the readability changes. 

We discussed off-list and decided to implement the remaining comments I raised, but to postpone them to a later pull request so that this can be available for the 0.1% survey if it begins over the weekend.
